tests/helpers/containers.py

```diff
@@ -0,0 +1,15 @@
+__all__ = ["expvar"]
+
+import pytest
+
+from tests.helpers.assertions import tcp_socket_open
+from tests.helpers.util import run_service, container_ip, wait_for
+
+
+@pytest.yield_fixture
+def expvar():
```

Comment:
This is going to be cleaner and more scalable if it just goes in the package with the expvar monitor tests.  If this file actually had all the containers, it would be gigantic.

Also consider this, if you just have a function like this:

```
def run_expvar(...):
    with run_service("expvar") as expvar_container:
        host = container_ip(expvar_container)
        assert wait_for(lambda: tcp_socket_open(host, 8080), 60), "service didn't start"
        yield host
```

Then you can wrap it as a fixture if you really want to by just doing:

```
@pytest.yield_fixture
def expvar():
    with run_expvar() as host:
        yield host
```

It is much clearer though to see in a test `with run_expvar(...) as host` and know "this runs expvar" than it is to see some magic function parameter that seems to come out of nowhere and there wouldn't even be a clear import in that same file so somebody not intimately familiar with the tests has to grep through the test dir to find that the fixture comes from a central python file in a very indirect way.

Doing `with run_expvar(...) as host` also lets you customize the running of the container (this isn't that uncommon in the existing tests), which fixtures don't apparently allow from the test function.  It is one line longer and one indentation level deeper to use a context manager instead of a fixture, but it is clearer and gives you way more control over the start/stop of the service.

At the very least, please don't shove the core logic of this stuff in a fixture function where it is totally unreusable, but it in a separate context manager that is reusable directly from the tests and then make a wrapping fixture if there is a super simple test case that wants to use fixtures to save a line.

Reply:
You can pass parameters to fixtures using [fixture factories](https://docs.pytest.org/en/latest/fixture.html#factories-as-fixtures). I'll remove the central containers.py and move it into `tests/monitors/expvar/conftest.py`. It could go in the test file itself but then you get name shadowing that makes pylint unhappy. So you do know you have to look in the tests own `conftest.py` but it's localized in the tests own package. I think that should alleviate most concerns.

Reply:
If you use fixture factories, it is just a more convoluted way of doing `with run_expvar(...) as host` since you will still have to make it a context manager to clean up the container.  The only thing a fixture factory would gain in that case is not having to have an explicit import, but it also makes it less clear to developers what is going on, which is my biggest concern about all of the indirection.

Reply:
The fixture factory doesn't require using a `with` in the test because the fixture handles the teardown part itself. But I agree it doesn't buy you much. I think for cases where we don't need anything special from the fixture and just want a container it's pretty straightforward. Perhaps naming the fixture `expvar_container` instead would make it more explicit. See what you think in the next push.

I agree that there is some magic to it given the nature of pytest fixtures but if all tests adhere to the convention then I don't think it'll negatively impact clarity.

Reply:
Ok, if the factory thing works without having to have a `with` block and extra indentation then it might be worth it to use.  Especially if it allows you to control container start/stop in the test function then it would allow for more advanced tests without having to have a distinct pattern for doing containers.